oauth-common/src/main/java/io/strimzi/kafka/oauth/jsonpath/JsonPathFilterQuery.java

```diff
@@ -0,0 +1,456 @@
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.kafka.oauth.jsonpath;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import java.math.BigDecimal;
+import java.util.ArrayList;
+
+/**
+ * This class implements parsing JsonPath filter query syntax inspired by:
+ *
+ *    https://github.com/json-path/JsonPath
+ *
+ * Given the following content of the JWT token:
+ * <pre>
+ *   {
+ *     "aud": ["uma_authorization", "kafka"],
+ *     "iss": "https://keycloak/token/",
+ *     "iat": 0,
+ *     "exp": 600,
+ *     "sub": "username",
+ *     "custom": "custom-value",
+ *     "roles": {
+ *       "client-roles": {
+ *         "kafka": ["kafka-user"]
+ *       }
+ *     },
+ *     "custom-level": 9
+ *   }
+ * </pre>
+ *
+ * Some examples of valid queries are:
+ *
+ * <pre>
+ *   "@.exp &lt; 1000"
+ *   "@.custom == 'custom-value'"
+ *   "@.custom == 'custom-value' and @.exp &gt; 1000"
+ *   "@.custom == 'custom-value' or @.exp &gt;= 1000"
+ *   "@.custom == 'custom-value' &amp;&amp; @.exp &lt;= 1000"
+ *   "@.custom != 'custom-value'"
+ *   "@.iat != null"
+ *   "@.iat == null"
+ *   "@.custom in ['some-custom-value', 42, 'custom-value']"
+ *   "@.custom nin ['some-custom-value', 42, 'custom-value']"
+ *   "@.custom-level in [1,8,9,20]"
+ *   "@.custom-level nin [1,2,3]"
+ *   "@.roles.client-roles.kafka != null"
+ *   "'kafka' in @.aud"
+ *   '"kafka-user" in @.roles.client-roles.kafka'
+ *   "@.exp &gt; 1000 || 'kafka' in @.aud"
+ * </pre>
+ *
+ * This class only implements a subset of the JsonPath syntax. It is focused on filter matching - answering the question if the JWT token matches the selector or not.
+ * <p>
+ * Main difference with the JsonPath is that the attribute paths using '@' match relative to root JSON object rather than any child attribute.
+ * For equivalent queries using other JsonPath implementations one would have to wrap the JWT object into another attribute, for example:
+ * <pre>
+ *   {
+ *       "token": {
+ *         "sub": "username",
+ *         ...
+ *       }
+ *   }
+ * </pre>
+ * and perform queries of the format:
+ * <pre>
+ *    $[*][?(QUERY)]
+ * </pre>
+ * For example: '$[*][?(@.sub != null &amp;&amp; )]'
+ *
+ * Some other differences are:
+ * <ul>
+ * <li> the use of 'or' / 'and' in addition to '||' / '&amp;&amp;'</li>
+ * <li> the requirement to use whitespace between operands and operators</li>
+ * </ul>
+ *
+ * Usage:
+ * <pre>
+ *   JsonPathFilterQuery query = new JsonPathFilterQuery("@.custom == 'value'");
+ *   boolean match = query.match(jsonObject);
+ * </pre>
+ *
+ * Query is parsed in the first line and any errors during parsing result
+ * in {@link JsonPathFilterQueryException}.
+ *
+ * Matching is thread safe. The normal usage pattern is to initialise the JsonPathFilterQuery object once,
+ * and query it many times concurrently against json objects.
+ */
+public class JsonPathFilterQuery {
```

Comment:
If you're going to write your own parser like this it would be really useful to write down the grammar of the language you're trying to parse. It makes it much easier to read each of the methods to decide whether they're doing the right thing, and it also makes a bit more formal the sorts of input you're accepting.

Reply:
Even better - just use a third party lib, but I didn't find any satisfactory one. Ideally I can just point to some third party library grammar, like in this case I point to the json-path/JsonPath project ...

My idea was that the number of expressions tested in JsonPathFilterQueryTest.java should serve as a source of truth or the 'language spec'.